src/tactic/sort.lean

```diff
@@ -0,0 +1,156 @@
+/-
+Copyright (c) 2022 Arthur Paulino, Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Arthur Paulino, Damiano Testa
+-/
+import data.polynomial.degree.definitions
+
+/-!  # A tactic for sorting sums  -/
+
+namespace tactic
+
+/--  Given an expression `ei` and a list `l : list expr`, returns a `tactic expr` that is
+`ei + sum_of_list l`. -/
+meta def build_sum : expr → list expr → tactic expr
+| ei (e :: es) := do
+  e' ← build_sum ei es,
+  mk_app `has_add.add [e', e]
+| ei [] := pure ei
+
+/--  Takes an `expr` and returns a list of its summands. -/
+meta def get_summands : expr → list expr
+| `(%%a + %%b) := get_summands a ++ get_summands b
+| a            := [a]
+
+section with_weights
+variables {N : Type*} [has_lt N] [decidable_rel (has_lt.lt : N → N → Prop)]
+
+/--  Given an expression `a` and a weight function `wt : expr → N` to a Type `N` with `<`,
+`sort_summands_with_weight a wt` returns the list of summands appearing in `a`, sorted using the
+order `<`. -/
+meta def sort_summands_with_weight : expr → (expr → N) → list expr
+| a wt := (get_summands a).qsort (λ e f, wt e < wt f)
```

Comment:
Could spell this
```suggestion
| a wt := (get_summands a).qsort $ (<) on wt
```
I'd be inclined to move a and wt before the colon too

Reply:
I moved the arguments before the colon.

I tried using `on` or `on_fun` or `function.on_fun`, but I think that there is an issue with the fact that `(<)` has Type `N → N → Prop` instead of `N → N → bool`.

Reply:
Weird, `merge_sort` takes a Prop relation, but quicksort doesn't

Reply:
Doing
```
meta def sort_summands_with_weight (a : expr) (wt : expr → N) : list expr :=
(get_summands a).qsort $ function.on_fun (<) wt
```
says:
```
type mismatch at application
  function.on_fun has_lt.lt
term
  has_lt.lt
has type
  ?m_1 → ?m_1 → Prop : Type ?
but is expected to have type
  ?m_1 → ?m_1 → bool : Sort (max ? 1)
```

Reply:
And even with a type annotation:
```
(get_summands a).qsort $ function.on_fun (has_lt.lt : N → N → Prop) wt
```
says
```
type mismatch at application
  function.on_fun has_lt.lt
term
  has_lt.lt
has type
  N → N → Prop : Type u_1
but is expected to have type
  ?m_1 → ?m_1 → bool : Sort (max ? 1)
```

Reply:
I'm wondering if, instead of a type with `<`, maybe it would be simpler to pass a function `N →N →bool`, use the notation that Eric suggests and then apply this with `has_lt.lt`.

Any opinion?

Reply:
Did you miss my comment above about `merge_sort`? That has the nicer interface

Reply:
I misunderstood your comment: with `merge_sort`, everything is fine!